### PR TITLE
add target_location_assign

### LINKED ISSUE #7843: Add target_location_assign_op
Add target_location_assign_op
described in https://github.com/PaddlePaddle/Paddle/issues/7402 : target_assign_op

### PR DESCRIPTION
**Pull request by wanghaox**
resolve https://github.com/PaddlePaddle/Paddle/issues/7843

### CODE REVIEW

**Review by qingqing01 — COMMENTED** [review]
Only part code review,  others will review again.

File: python/paddle/v2/fluid/tests/test_target_location_assign_op.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        # match_indices[n, p] = gt_box_index
+        self.match_indices = np.zeros(
+            (self.batch_size, self.prior_num)).astype('int32')
+
+        for n in range(self.batch_size):
+            gt_start = self.encode_gt_boxes_lod[0][n]
+            gt_end = self.encode_gt_boxes_lod[0][n + 1]
+            gt_num = gt_end - gt_start
+            for p in range(self.prior_num):
+                self.loc_predic[n, p, :] = make_box()
```
Comment (qingqing01) [inline python/paddle/v2/fluid/tests/test_target_location_assign_op.py]: In fact, `loc_predic` is the localization prediction offsets, not bbox. So using `random.random` is ok.

File: python/paddle/v2/fluid/tests/test_target_location_assign_op.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+                gt_encode = self.encode_box(self.prior_boxes[j],
+                                            self.prior_variance[j], gt_box,
+                                            encode_variance_in_target)
+
+                loc_pred = self.loc_predic[i, j, :]
+                if (encode_variance_in_target):
+                    for k in range(4):
+                        loc_pred[k] /= self.prior_variance[j, k]
+                        gt_encode[k] /= self.prior_variance[j, k]
```
Comment (qingqing01) [inline python/paddle/v2/fluid/tests/test_target_location_assign_op.py]: Remove for loop:

```python
loc_pred = self.loc_predic[i, j, :] /self.prior_variance[j, :]
gt_encode = gt_encode/self.prior_variance[j, :]
```

尽量使用 Pythonic的代码，能向量操作的，向量操作。

File: paddle/operators/target_location_assign_op.cc [hunk 3]
```diff
@@ -0,0 +1,150 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
```
[the remark this answers is not in the document]
Comment (wanghaox) [reply paddle/operators/target_location_assign_op.cc]: done

File: paddle/operators/target_location_assign_op.cc [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    auto mi_dims = ctx->GetInputDim("MatchIndices");
+    auto pb_dims = ctx->GetInputDim("PriorBoxes");
+    auto pv_dims = ctx->GetInputDim("PriorVariances");
+    int prior_box_num = pb_dims[0];
+    PADDLE_ENFORCE_EQ(loc_dims.size(), 3UL,
+                      "The rank of Input(Loc) must be 3, the shape is "
+                      "[batch_size, prior_box_num, 4].");
+    PADDLE_ENFORCE_EQ(gt_dims.size(), 3UL,
+                      "The rank of Input(GTBoxes) must be 3, the shape is "
+                      "[N, prior_box_num, 4].");
```
[the remark this answers is not in the document]
Comment (wanghaox) [reply paddle/operators/target_location_assign_op.cc]: done

File: paddle/operators/target_location_assign_op.cc [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    PADDLE_ENFORCE_EQ(mi_dims.size(), 2UL,
+                      "The rank of Input(MatchIndices) must be 2, the shape is "
+                      "[batch_size, prior_box_num].");
+    PADDLE_ENFORCE_EQ(pb_dims.size(), 2UL,
+                      "The rank of Input(PriorBoxes) must be 2, the shape is "
+                      "[prior_box_num, 4].");
+    PADDLE_ENFORCE_EQ(
+        pv_dims.size(), 2UL,
+        "The rank of Input(PriorVariances) must be 2, the shape is "
+        "[prior_box_num, 4].");
```
[the remark this answers is not in the document]
Comment (wanghaox) [reply paddle/operators/target_location_assign_op.cc]: done

**Pull request closed without merging**